### ISSUE TITLE
Error processing a message with key NULL.

### ISSUE THREAD

**Issue by hclbaur** [Axual/ksml#47]
I am not 100% sure what is causing this error in the ksml/runner and also not why it is referring to a Json deserialiser:

`Caused by: io.axual.ksml.exception.KSMLExecutionException: Error while executing the Kafka Streams topology: Json serde not found for data type null`

I believe it has to do with the fact that the message has no key (or a key with a null value). 
The message looks like this (in message browser in the Axual Self-service environment):

````
"message":{
"partition":4
"offset":951
"timestamp":1652912264843
"headers":[...]
"value":
"0000000911020e43524541544544021a4d6574657252656164696e6773021470352d70..."
"key":NULL
}
````

The key is NULL and the value is a binary string.  I configured the KSML app like this, with keyType none and valueType bytes:

````
streams:
  p5meterresponse:
    topic: APP9329_LianderP5MeterResponse
    keyType: none
    valueType: bytes

functions:
  print_message:
    type: forEach
    code: "print('key='+(key if isinstance(key,str) else str(key))+', value='+(value if isinstance(value,str) else str(value)))"

pipelines:
  main:
    from: p5meterresponse
    forEach: print_message
````

I also tried keyType string but then I see a null pointer exception when reading the message.

The message we are trying to read is actually an Avro message from the Liander platform, but for demo purpose I would like to read it as a binary type.

**Comment by tonvanbart** [comment]
Hi Harold, is there a stacktrace somewhere in the log and if so, could you post it? Thanks!

**Comment by hclbaur** [comment]
Sure, here it is.

[stacktrace.txt](https://github.com/Axual/ksml/files/8812852/stacktrace.txt)

Like I said, I think the issue may be caused by the key/value types in my configuration, except that I wouldn't know what type would be correct in this case. But maybe it is actually a bug?

**Comment by jeroenvandisseldorp** [comment]
We updated some of the code involved and introduced a better way of handling type "none". New images were pushed. In the future the solution may need some more refinement, but hopefully this solves the crashes for now. Can you pull and check?

**Comment by hclbaur** [comment]
In order to isolate the issue I have set up a personal test stream (test_harold) in the Axual environment and I tried to crash the runner by producing messages with and without a key. And sure enough omitting the key is enough to really upset it :)
This was with key type "string" by the way.

I will pull the new image, redeploy and try again... will let you know asap.

**Comment by jeroenvandisseldorp** [comment]
Thanks for verifying and good to hear!

If you know that the key is always left null, you could just specify "none" as the key type. That should work now in the new build equally well and will prevent parsing in any non-blank (ie. non-string) key that the other party may produce on the topic.

On the side note: consumer offsets should be committed automatically. Crashes of the KSML runtime may cause offsets not completely committed but this should not occur for normal runs. Did you verify this restarting of consumption on the new build? Are you using the same consumer group id / application id between runs?

**Comment by tonvanbart** [comment]
Closed by #48 .

**Comment by hclbaur** [comment]
OK, I re-tested. First of all, the messages are committed just fine. But I noticed that sometimes messages seem to be processed after quite a long delay, which may have confused me. Nevertheless, I seem to get them only once between restarts.

As for the original issue, that appears to be solved. Specifying either key type `string` or `none ` will work without crashing; the difference being that for key type `none` the key is actually ignored and logged as "None". This is a nice one for the documentation btw, as type `none` seems to be an undocumented feature. I found it by looking at the source code :)

Thanks for fixing this issue, I will close it.
